### DISCUSSION TITLE
Support generating absolute asset paths

### DISCUSSION THREAD

**Discussion by zadeviggers** [withastro/roadmap#79] [Proposal]
I would like it if Astro supported referencing generated assets (css and the like) with absolute paths rather than relative paths (which is what it currently does).
My main motivation is an issue I'm having with 404 pages not loading their css properly because the relative path is being generated incorrectly (although I think there may already be an open github issue about this), but I also generally prefer it because I feel that it makes the site source easier to understand.
I understand the use case where relative asset paths are important (when the generated site is not hosted at the root), which is why I think this should be added as a build configuration option, and disabled by default - perhaps `buildOptions.assetPathType: "relative" | "absolute" = "relative"` or `buildOptions.absoluteAssetPaths: boolean = false`.

Example of relative generated asset paths (current):
```html
<link rel="stylesheet" type="text/css" href="../assets/globals.c2ce62df.css">
<link rel="stylesheet" type="text/css" href="../assets/index.fccc3c9f.css">
```
Example of absolute generated asset paths (proposed option):
```html
<link rel="stylesheet" type="text/css" href="/assets/globals.c2ce62df.css">
<link rel="stylesheet" type="text/css" href="/assets/index.fccc3c9f.css">
```

**Comment by jonathantneal** [comment]
I would expand upon this proposal to target more types of links with more options, i.e:

```js
interface BuildOptionsProposal {
  linkFormat: "relative" | "root-relative" | "domain-relative" | "absolute"
}
```

This option would produce HTML as follows:

**relative**

```html
<a href="sibling-page">Sibling Page</a>
```

**root-relative**

```html
<a href="/example-subdir/sibling-page">Sibling Page</a>
```

**domain-relative**

```html
<a href="//example.com/example-subdir/sibling-page">Sibling Page</a>
```

**absolute**

```html
<a href="https://example.com/example-subdir/sibling-page">Sibling Page</a>
```

**Reply by zadeviggers** [reply]
I like this idea, but I think that the option name should be something more general than `linkFormat`.

**Comment by jonathantneal** [comment]
### Early Feedback from the Community RFC Call (2022-01-25)

- No consensus. Mixed opinions.
- 1 person expressed support for this, and other link formats.
- 3 people expressed dropping support for **relative** links altogether and only supporting **root-relative**.
- 1 person mentioned their current use case for the current **relative** behavior.
- Recommendation to @withastro/maintainers-core: If link formatting is subject to change, write up an RFC.

**Reply by zadeviggers** [reply]
Sorry that I couldn't attend the RFC call.

What if we supported all of the proposed formats and had it set to root-relative by default?

**Reply by matthewp** [reply]
One big issue we had was the maintenance cost of supporting all of these options. We already have 2 related config options `trailingSlash` subpath support and neither are supported particularly well.

Another issue is that `relative` is not feasible because Vite uses non-relative paths is some of the code it produces. So we would need to upstream a change there.

**Reply by jonathantneal** [reply]
> Another issue is that `relative` is not feasible because Vite uses non-relative paths is some of the code it produces

That seems like a bug. What’s the non-relative path that Vite needs? It will be so hard to re-introduce `relative` if we lose it.

**Comment by jonathantneal** [comment]
We have an RFC call tomorrow. If link formatting is subject to change, I can add this as an agenda item. As the newly suggested change came from core and not the original author, let me know if you would need help from the community or myself to write it up.

**Reply by zadeviggers** [reply]
Yes, I'd appreciate this being in tomorrow's rfc call agenda. I'm not really sure how to go about writing it up so some help would be appreciated.
